File: src/git_slop/reports/sarif.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from git_slop.reports.shared import (
    EXPECTED_REPORT_SCHEMA_VERSION,
    resolve_path,
    strongest_pressures,
)

SARIF_SCHEMA_VERSION = 1
SARIF_VERSION = "2.1.0"
SARIF_SCHEMA_URL = "https://json.schemastore.org/sarif-2.1.0.json"
BOUNDARY_NOTE = (
    "SARIF export boundary: this is a deterministic projection of existing "
    "git-slop report evidence. It does not rerun the detector, upload results, "
    "mutate code, or change detector scoring semantics."
)
RULE_ID = "git-slop.hotspot"


def _validate_report(report: dict[str, Any]) -> None:
    if report.get("schema_version") != EXPECTED_REPORT_SCHEMA_VERSION:
        raise ValueError(f"git slop sarif requires report schema {EXPECTED_REPORT_SCHEMA_VERSION}.")


def _level_for_record(record: dict[str, Any]) -> str:
    priority_band = record.get("priority_band")
    context_band = record.get("context_band")
    if priority_band == "must_refactor" or context_band == "critical":
        return "error"
    if priority_band in {"should_refactor", "needs_refactor"} or context_band == "warning":
        return "warning"
    return "note"


def _record_for_queue_item(report: dict[str, Any], item: dict[str, Any]) -> dict[str, Any]:
    record = resolve_path(report, item["path"])
    if record is None:
        return dict(item)
    payload = dict(record)
    payload.setdefault("priority_score", item.get("priority_score"))
    payload.setdefault("priority_band", item.get("priority_band"))
    payload.setdefault("context_band", item.get("context_band"))
    payload.setdefault("reason_codes", item.get("reason_codes", []))
    return payload


def _message_for_record(record: dict[str, Any]) -> str:
    reason_codes = record.get("reason_codes") or []
    reasons = ", ".join(reason_codes) if reason_codes else "no reason codes"
    return (
        f"{record.get('path')} is ranked {record.get('priority_band')} "
        f"with score {record.get('priority_score')} and context "
        f"{record.get('context_band')} ({reasons})."
    )


def _overlay_properties(record: dict[str, Any]) -> dict[str, float]:
    return {
        label: round(value, 6)
        for label, value in strongest_pressures(record.get("overlays", {}) or {}, limit=8)
    }


def _result_for_record(record: dict[str, Any], *, rank: int) -> dict[str, Any]:
    path = str(record.get("path"))
    return {
        "ruleId": RULE_ID,
        "ruleIndex": 0,
        "level": _level_for_record(record),
        "message": {"text": _message_for_record(record)},
        "locations": [
            {
                "physicalLocation": {
                    "artifactLocation": {"uri": path},
                },
            }
        ],
        "properties": {
            "git_slop": {
                "rank": rank,
                "priority_score": record.get("priority_score"),
                "priority_band": record.get("priority_band"),
                "context_band": record.get("context_band"),
                "reason_codes": record.get("reason_codes", []),
                "costs": record.get("costs", {}),
                "strongest_overlays": _overlay_properties(record),
                "evidence_boundary": (
                    "Hotspot cost and overlay evidence are preserved as separate "
                    "properties; SARIF export does not rescore the finding."
                ),
            }
        },
    }


def _rules() -> list[dict[str, Any]]:
    return [
        {
            "id": RULE_ID,
            "name": "Git Slop hotspot",
            "shortDescription": {
                "text": "File ranked in the git-slop action queue.",
            },
            "fullDescription": {
                "text": (
                    "A deterministic git-slop hotspot based on context cost. "
                    "Overlay evidence is exported separately in result properties "
                    "and does not change detector scoring."
                ),
            },
            "help": {
                "text": (
                    "Review the git-slop report, explain output, or plan output for "
                    "supporting evidence before deciding whether maintenance work is "
                    "appropriate."
                ),
            },
            "properties": {
                "precision": "medium",
                "tags": ["maintainability", "context-cost", "git-slop"],
            },
        }
    ]


def build_sarif_payload(
    report: dict[str, Any],
    *,
    report_path: str | None = None,
    top: int | None = None,
) -> dict[str, Any]:
    _validate_report(report)
    if top is not None and top <= 0:
        raise ValueError("--top must be greater than zero.")

    action_queue = report.get("action_queue", [])
    selected = action_queue[:top] if top is not None else action_queue
    results = [
        _result_for_record(_record_for_queue_item(report, item), rank=index)
        for index, item in enumerate(selected, start=1)
        if isinstance(item.get("path"), str)
    ]
    repo = report.get("repo", {}) or {}
    return {
        "$schema": SARIF_SCHEMA_URL,
        "version": SARIF_VERSION,
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": "git-slop",
                        "informationUri": "https://github.com/coreycoto/git-slop",
                        "rules": _rules(),
                    }
                },
                "automationDetails": {
                    "id": "git-slop/sarif",
                },
                "versionControlProvenance": [
                    {
                        "repositoryUri": repo.get("remote_url") or repo.get("repo_name"),
                        "revisionId": repo.get("head_sha"),
                    }
                ],
                "invocations": [
                    {
                        "executionSuccessful": True,
                        "properties": {
                            "git_slop": {
                                "schema_version": SARIF_SCHEMA_VERSION,
                                "report_schema_version": report.get("schema_version"),
                                "report_path": report_path,
                                "boundary_note": BOUNDARY_NOTE,
                            }
                        },
                    }
                ],
                "results": results,
                "properties": {
                    "git_slop": {
                        "summary": report.get("summary", {}),
                        "stats": report.get("stats", {}),
                        "boundary_note": BOUNDARY_NOTE,
                    }
                },
            }
        ],
    }


def render_sarif_json(payload: dict[str, Any]) -> str:
    return json.dumps(payload, indent=2, sort_keys=True) + "\n"


def write_sarif_file(payload: dict[str, Any], output_path: Path) -> None:
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(render_sarif_json(payload), encoding="utf-8")

File: src/report_ops/sarif.rs
use super::*;
use sha2::{Digest, Sha256};

fn sarif_level(record: &Value) -> &'static str {
    let slop_band = string(record.get("slop_band"));
    let context_band = string(record.get("context_band"));
    if slop_band == "critical"
        || matches!(
            context_band.as_str(),
            "critical" | "refactor_required" | "budget_exceeded"
        )
    {
        "error"
    } else if matches!(slop_band.as_str(), "high" | "moderate") || context_band == "warning" {
        "warning"
    } else {
        "note"
    }
}

fn sarif_record(report: &Value, queue_item: &Value) -> Value {
    let path = string(queue_item.get("path"));
    let mut record = find_record(report, &path)
        .map(|(record, _)| record)
        .unwrap_or_else(|| queue_item.clone());
    let record_object = record.as_object_mut().expect("record object");
    for key in [
        "classification",
        "remediation_kind",
        "slop_score",
        "slop_band",
        "context_band",
        "reason_codes",
    ] {
        if !record_object.contains_key(key) {
            record_object.insert(
                key.to_string(),
                queue_item.get(key).cloned().unwrap_or_else(|| {
                    if key == "reason_codes" {
                        json!([])
                    } else {
                        Value::Null
                    }
                }),
            );
        }
    }
    record
}

fn sarif_result(record: &Value, rank: usize) -> Value {
    let path = string(record.get("path"));
    let reasons = string_array(record.get("reason_codes"));
    let reasons_text = if reasons.is_empty() {
        "no reason codes".to_string()
    } else {
        reasons.join(", ")
    };
    let overlays: Map<String, Value> = strongest_pressures(record.get("overlays"), 8)
        .into_iter()
        .map(|(label, value)| (label, json!(round6(value))))
        .collect();
    let context_finding = matches!(
        string(record.get("context_band")).as_str(),
        "warning" | "critical" | "refactor_required" | "budget_exceeded"
    );
    let rule_id = if context_finding {
        "git-slop.context-budget"
    } else {
        "git-slop.maintenance-pressure"
    };
    let fingerprint = hex::encode(Sha256::digest(format!("{rule_id}\0{path}").as_bytes()));
    let mut result = json!({
        "ruleId": rule_id,
        "ruleIndex": if context_finding { 0 } else { 1 },
        "level": sarif_level(record),
        "message": {
            "text": format!(
                "{path} is ranked {} with slop_score {} and context {} ({reasons_text}).",
                json_scalar_text(record.get("slop_band")),
                json_scalar_text(record.get("slop_score")),
                json_scalar_text(record.get("context_band")),
            ),
        },
        "locations": [{
            "physicalLocation": {
                "artifactLocation": {"uri": path},
            },
        }],
        "partialFingerprints": {"gitSlopFinding/v1": fingerprint},
        "properties": {
            "git_slop": {
                "rank": rank,
                "slop_score": record.get("slop_score").cloned().unwrap_or(Value::Null),
                "slop_band": record.get("slop_band").cloned().unwrap_or(Value::Null),
                "context_band": record.get("context_band").cloned().unwrap_or(Value::Null),
                "classification": record.get("classification").cloned().unwrap_or(Value::Null),
                "remediation_kind": record.get("remediation_kind").cloned().unwrap_or(Value::Null),
                "reason_codes": record.get("reason_codes").cloned().unwrap_or_else(|| json!([])),
                "costs": record.get("costs").cloned().unwrap_or_else(|| json!({})),
                "strongest_overlays": Value::Object(overlays),
                "evidence_boundary": "Hotspot cost and overlay evidence are preserved as separate properties; SARIF export does not rescore the finding.",
            },
        },
    });
    if let Some(state) = record.get("baseline_state").and_then(Value::as_str) {
        result["baselineState"] = json!(state);
    }
    result
}

pub fn sarif_payload(
    report: &Value,
    report_path: Option<&str>,
    top: Option<usize>,
    scope: &str,
) -> Result<Value> {
    require_report_schema(report, "sarif")?;
    if top == Some(0) {
        bail!("--top must be greater than zero.");
    }
    let policy_records;
    let queue = if scope == "policy" {
        let context = report
            .pointer("/config/check/fail_on_context_band")
            .and_then(Value::as_str);
        let slop = report
            .pointer("/config/check/fail_on_slop_band")
            .and_then(Value::as_str);
        policy_records = super::failing_records_in(report, context, slop, false);
        policy_records.as_slice()
    } else if scope == "action-queue" {
        array_at(report, &["action_queue"])
    } else {
        bail!("SARIF scope must be policy or action-queue.");
    };
    let take = top.unwrap_or(queue.len());
    let results: Vec<Value> = queue
        .iter()
        .take(take)
        .enumerate()
        .filter_map(|(index, item)| {
            item.get("path")
                .and_then(Value::as_str)
                .map(|_| sarif_result(&sarif_record(report, item), index + 1))
        })
        .collect();
    let returned = results.len();
    let help_uri = format!(
        "https://github.com/coreycoto/git-slop/blob/v{}/docs/scoring-model.md",
        crate::VERSION
    );
    let repo = report.get("repo").unwrap_or(&Value::Null);
    let repository_uri = ["remote_url", "git_remote_url", "repo_name"]
        .into_iter()
        .find_map(|key| {
            repo.get(key)
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
        })
        .map_or(Value::Null, |value| json!(value));
    let revision_id = ["head_sha", "head_commit"]
        .into_iter()
        .find_map(|key| {
            repo.get(key)
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
        })
        .map_or(Value::Null, |value| json!(value));
    Ok(json!({
        "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
        "version": "2.1.0",
        "runs": [{
            "tool": {
                "driver": {
                    "name": "git-slop",
                    "informationUri": "https://github.com/coreycoto/git-slop",
                    "rules": [{
                        "id": "git-slop.context-budget",
                        "name": "Git Slop context budget",
                        "shortDescription": {"text": "File meets a configured context-cost threshold."},
                        "fullDescription": {"text": "A deterministic context-cost finding from Git Slop."},
                        "helpUri": help_uri,
                        "help": {"text": "Review the git-slop report, explain output, or plan output for supporting evidence before deciding whether maintenance work is appropriate."},
                        "properties": {
                            "precision": "medium",
                            "tags": ["maintainability", "context-cost", "git-slop"],
                        },
                    }, {
                        "id": "git-slop.maintenance-pressure",
                        "name": "Git Slop maintenance pressure",
                        "shortDescription": {"text": "File meets a configured maintenance-pressure threshold."},
                        "fullDescription": {"text": "A deterministic maintenance-pressure finding from Git Slop."},
                        "helpUri": help_uri,
                        "properties": {"precision": "medium", "tags": ["maintainability", "git-slop"]}
                    }],
                },
            },
            "automationDetails": {"id": "git-slop/sarif"},
            "versionControlProvenance": [{
                "repositoryUri": repository_uri,
                "revisionId": revision_id,
            }],
            "invocations": [{
                "executionSuccessful": true,
                "properties": {
                    "git_slop": {
                        "schema_version": SARIF_SCHEMA_VERSION,
                        "report_schema_version": report.get("schema_version").cloned().unwrap_or(Value::Null),
                        "report_path": report_path,
                        "analyzer": report.get("analyzer").cloned().unwrap_or_else(|| json!({})),
                        "boundary_note": SARIF_BOUNDARY_NOTE,
                        "scope": scope,
                        "collection": {
                            "total": queue.len(),
                            "returned": returned,
                            "limit": top,
                            "truncated": returned < queue.len()
                        },
                    },
                },
            }],
            "results": results,
            "properties": {
                "git_slop": {
                    "summary": report.get("summary").cloned().unwrap_or_else(|| json!({})),
                    "stats": report.get("stats").cloned().unwrap_or_else(|| json!({})),
                    "boundary_note": SARIF_BOUNDARY_NOTE,
                },
            },
        }],
    }))
}

pub fn render_json(payload: &Value) -> Result<String> {
    Ok(format!("{}\n", serde_json::to_string_pretty(payload)?))
}
